In [ ]:
import sys
from pathlib import Path

PROJECT_ROOT = Path().resolve().parent
sys.path.insert(0, str(PROJECT_ROOT))

from src.data.load_data import load_bank_data
from src.data.preprocess import split_data, BankFeatureEngineer, build_enhanced_pipeline
from src.models.knn import (
    train_knn,
    find_best_threshold,
    evaluate_knn,
    save_knn,
)

In [ ]:
#training KNN block

# Load dataset
df = load_bank_data("../data/raw/bank-full.csv")

# Split
X_train, X_test, y_train, y_test = split_data(df)

# Detect engineered columns
fe = BankFeatureEngineer()
X_tmp = fe.fit_transform(X_train)

num_cols = X_tmp.select_dtypes(include=["int64", "float64"]).columns.tolist()
cat_cols = X_tmp.select_dtypes(include=["object", "category"]).columns.tolist()

# Build preprocessing pipeline
preprocess_pipeline = build_enhanced_pipeline(num_cols, cat_cols)

# Train KNN
best_knn, best_params, best_cv_f1 = train_knn(
    preprocess_pipeline,
    X_train,
    y_train
)

print("Best params:", best_params)
print("Best CV F1:", best_cv_f1)

Fitting 5 folds for each of 12 candidates, totalling 60 fits
Best params: {'classifier__metric': 'manhattan', 'classifier__n_neighbors': 10, 'classifier__weights': 'distance'}
Best CV F1: 0.27511202771483456


In [ ]:
#Evaluation block

# Find optimal threshold
best_threshold, best_f1 = find_best_threshold(
    best_knn,
    X_test,
    y_test
)

print("Best threshold:", best_threshold)
print("Best F1:", best_f1)

# Final evaluation
metrics = evaluate_knn(
    best_knn,
    X_test,
    y_test,
    best_threshold
)

print("\n--- FINAL KNN ---")
for k, v in metrics.items():
    print(f"{k}: {v}")

# Save model
save_knn(best_knn, best_threshold)


Best threshold: 0.25747474747474747
Best F1: 0.40857273141814865

--- FINAL KNN ---
Accuracy: 0.8565741457480924
AUC: 0.7349823570423276
Precision: 0.3947136563876652
Recall: 0.42344045368620037
F1: 0.40857273141814865
MCC: 0.3273402686221143


### KNN Observations

- The dataset is highly imbalanced (~88% 'no', ~12% 'yes'), so F1-score is used for evaluation.
- KNN does not support class weighting, so imbalance is handled indirectly using distance-based weighting and threshold tuning.

- Hyperparameter tuning using GridSearchCV selected:
  → n_neighbors = 10  
  → metric = Manhattan  
  → weights = distance  

- Feature scaling using StandardScaler is essential for KNN, as it relies on distance calculations.

- AUC score (~0.73) is lower compared to Logistic Regression and Decision Tree, indicating weaker class separation.

- Default threshold (0.5) is not optimal for imbalanced data.

- After threshold tuning, best threshold ≈ 0.26 significantly improves recall and F1-score.  
  → Lower threshold increases detection of positive class.

- Precision (~0.39) and Recall (~0.42) show moderate performance:
  → Model captures some positive cases  
  → But also introduces false positives  

- F1-score (~0.41) is lower than Decision Tree (~0.48), indicating weaker performance.

- MCC (~0.33) confirms lower overall predictive quality compared to other models.

**Conclusion:**  
KNN performs reasonably but is not well-suited for imbalanced datasets. It struggles to capture minority class patterns compared to Logistic Regression and Decision Tree.
